Crash when running in Android Studio 4.0.1

Reported by an anonymous contributor:

```
* Exception is:
org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreationException: Could not create task ':app:countDebugDexMethods'.
	at org.gradle.api.internal.tasks.DefaultTaskContainer.taskCreationException(DefaultTaskContainer.java:719)
	at org.gradle.api.internal.tasks.DefaultTaskContainer.access$600(DefaultTaskContainer.java:77)
	<...SNIP...>
Caused by: java.lang.IllegalStateException: Cannot call Task.finalizedBy(Object...) on task ':app:packageInternalDebug' after task has started execution.
	at org.gradle.api.internal.tasks.TaskMutator.mutate(TaskMutator.java:36)
	at org.gradle.api.internal.AbstractTask.finalizedBy(AbstractTask.java:819)
	at com.getkeepsafe.dexcount.LegacyTaskApplicator$createTask$1.execute(Plugin.kt:301)
	at com.getkeepsafe.dexcount.LegacyTaskApplicator$createTask$1.execute(Plugin.kt:164)
	at org.gradle.api.internal.DefaultMutationGuard$2.execute(DefaultMutationGuard.java:42)
	at org.gradle.api.internal.DefaultMutationGuard$2.execute(DefaultMutationGuard.java:42)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1$1.run(DefaultCollectionCallbackActionDecorator.java:100)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:60)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:97)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:395)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:387)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:84)
	at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction.execute(DefaultCollectionCallbackActionDecorator.java:94)
	at org.gradle.internal.ImmutableActionSet$SetWithManyActions.execute(ImmutableActionSet.java:329)
	at org.gradle.api.internal.DefaultDomainObjectCollection.doAdd(DefaultDomainObjectCollection.java:264)
	at org.gradle.api.internal.DefaultNamedDomainObjectCollection.doAdd(DefaultNamedDomainObjectCollection.java:113)
	at org.gradle.api.internal.DefaultDomainObjectCollection.add(DefaultDomainObjectCollection.java:258)
	at org.gradle.api.internal.DefaultNamedDomainObjectCollection$AbstractDomainObjectCreatingProvider.tryCreate(DefaultNamedDomainObjectCollection.java:944)
	... 129 more
```

This apparently only happened when running Android Studio - command-line builds worked just fine.  No idea how to reproduce this, but a valid workaround is disabling `runOnEachPackage`.  Maybe we should just rip that out entirely.

--- a/src/main/kotlin/com/getkeepsafe/dexcount/Plugin.kt
+++ b/src/main/kotlin/com/getkeepsafe/dexcount/Plugin.kt
@@ -197,9 +197,9 @@ abstract class LegacyTaskApplicator(ext: DexCountExtension, project: Project) : 
             else -> throw IllegalArgumentException("Dexcount plugin requires the Android plugin to be configured")
         }
 
-        variants.all { variant ->
+        variants.configureEach { variant ->
             if (!ext.enabled) {
-                return@all
+                return@configureEach
             }
 
             when (variant) {
@@ -256,7 +256,7 @@ abstract class LegacyTaskApplicator(ext: DexCountExtension, project: Project) : 
 
     protected fun createTask(
             variant: BaseVariant,
-            parentTask: Task,
+            parentTask: TaskProvider<*>,
             output: BaseVariantOutput?,
             applyInputConfiguration: (LegacyGeneratePackageTreeTask) -> Unit)  {
         var slug = variant.name.capitalize()
@@ -283,12 +283,9 @@ abstract class LegacyTaskApplicator(ext: DexCountExtension, project: Project) : 
             t.packageTreeFileProperty.set(project.layout.buildDirectory.file(treePath))
 
             applyInputConfiguration(t)
-
-            t.mustRunAfter(parentTask)
-            t.dependsOn(parentTask)
         }
 
-        project.tasks.register("count${slug}DexMethods", DexCountOutputTask::class.java) { t ->
+        val countTask = project.tasks.register("count${slug}DexMethods", DexCountOutputTask::class.java) { t ->
             t.description         = "Outputs dex method count for ${variant.name}."
             t.group               = "Reporting"
 
@@ -296,9 +293,11 @@ abstract class LegacyTaskApplicator(ext: DexCountExtension, project: Project) : 
             t.variantNameProperty.set(outputName)
             t.packageTreeFileProperty.set(gen.flatMap { it.packageTreeFileProperty })
             t.androidProject.set(true)
+        }
 
-            if (ext.runOnEachPackage) {
-                parentTask.finalizedBy(t)
+        if (ext.runOnEachPackage) {
+            parentTask.configure { t ->
+                t.finalizedBy(countTask)
             }
         }
     }
@@ -367,33 +366,22 @@ open class ThreeFourApplicator(ext: DexCountExtension, project: Project): Legacy
 
     override fun applyToLibraryVariant(variant: LibraryVariant) {
         val packageTaskProvider = variant.packageLibraryProvider
-        val packageTask = packageTaskProvider.orNull
-        if (packageTask == null) {
-            project.logger.error("LibraryVariant.getPackageLibraryProvider().getOrNull() unexpectedly returned null")
-            return
-        }
-
-        createTask(variant, packageTask, null) { t ->
-            t.inputFileProperty.set(packageTask.archiveFile)
+        createTask(variant, packageTaskProvider, null) { t ->
+            t.inputFileProperty.set(packageTaskProvider.flatMap { it.archiveFile })
         }
     }
 
     private fun applyToApkVariant(variant: ApkVariant) {
         checkPrintDeclarationsIsFalse()
 
-        variant.outputs.all { output ->
+        variant.outputs.configureEach { output ->
             if (output !is ApkVariantOutput) {
                 throw IllegalArgumentException("Unexpected output type for variant ${variant.name}: ${output::class.java}")
             }
 
-            val packageTask = variant.packageApplicationProvider.orNull
-            if (packageTask == null) {
-                project.logger.error("ApkVariant.getPackageApplicationProvider().getOrNull() unexpectedly returned null")
-                return@all
-            }
-
-            createTask(variant, packageTask, output) { t ->
-                val fileProvider = getOutputDirectory(packageTask).map { it.file(output.outputFileName) }
+            createTask(variant, variant.packageApplicationProvider, output) { t ->
+                val outputDirProvider = variant.packageApplicationProvider.flatMap { getOutputDirectory(it) }
+                val fileProvider = outputDirProvider.map { it.file(output.outputFileName) }
                 t.inputFileProperty.value(fileProvider)
             }
         }
@@ -434,7 +422,7 @@ open class FourOneApplicator(ext: DexCountExtension, project: Project) : Abstrac
             return
         }
 
-        project.plugins.withType(AppPlugin::class.java) {
+        project.plugins.withType(AppPlugin::class.java).configureEach {
             val android = project.extensions.getByType(ApplicationExtension::class.java)
             android.onVariantProperties {
                 registerApkTask()
@@ -442,14 +430,14 @@ open class FourOneApplicator(ext: DexCountExtension, project: Project) : Abstrac
             }
         }
 
-        project.plugins.withType(LibraryPlugin::class.java) {
+        project.plugins.withType(LibraryPlugin::class.java).configureEach() {
             val android = project.extensions.getByType(LibraryExtension::class.java)
             android.onVariantProperties {
                 registerAarTask(android.buildToolsRevision)
             }
         }
 
-        project.plugins.withType(TestPlugin::class.java) {
+        project.plugins.withType(TestPlugin::class.java).configureEach {
             val android = project.extensions.getByType(TestExtension::class.java)
             android.onVariantProperties {
                 registerApkTask()

